File: crates/pytest-rs-core/src/python/reporter.rs
//! Replacement terminalreporter delegation (sugar/pretty).

#[allow(unused_imports)]
use super::*;
use crate::collect::TestItem;
use crate::report::{Outcome, Phase};
use pyo3::types::PyDict;

static REPORTS_TEST_REPORT_CLS: pyo3::sync::PyOnceLock<Py<PyAny>> = pyo3::sync::PyOnceLock::new();
static REPORTER_LOGREPORT_FN: pyo3::sync::PyOnceLock<Py<PyAny>> = pyo3::sync::PyOnceLock::new();
static REPORTER_FEED_DEFAULT_FN: pyo3::sync::PyOnceLock<Py<PyAny>> = pyo3::sync::PyOnceLock::new();

fn reports_test_report_cls(py: Python<'_>) -> PyResult<Bound<'_, PyAny>> {
    REPORTS_TEST_REPORT_CLS
        .get_or_try_init(py, || {
            Ok(py
                .import("_pytest.reports")?
                .getattr("TestReport")?
                .unbind())
        })
        .map(|f| f.bind(py).clone())
}

fn reporter_logreport_fn(py: Python<'_>) -> PyResult<Bound<'_, PyAny>> {
    REPORTER_LOGREPORT_FN
        .get_or_try_init(py, || {
            Ok(py
                .import("pytest._reporter")?
                .getattr("logreport")?
                .unbind())
        })
        .map(|f| f.bind(py).clone())
}

fn reporter_feed_default_fn(py: Python<'_>) -> PyResult<Bound<'_, PyAny>> {
    REPORTER_FEED_DEFAULT_FN
        .get_or_try_init(py, || {
            Ok(py
                .import("pytest._reporter")?
                .getattr("feed_default")?
                .unbind())
        })
        .map(|f| f.bind(py).clone())
}

/// Register the default 'terminalreporter' plugin (the stand-in that
/// reporter-replacing plugins like pytest-sugar unregister). Must run
/// before python pytest_configure hooks fire.
pub fn reporter_setup(py: Python<'_>, config: &crate::config::Config) -> PyResult<()> {
    let config_proxy = make_py_config(py, config)?;
    py.import("pytest._reporter")?
        .getattr("setup")?
        .call1((config_proxy,))?;
    Ok(())
}

/// The plugin object that replaced 'terminalreporter' during configure,
/// or None when terminal output stays native.
pub fn reporter_replacement(py: Python<'_>) -> Option<Py<PyAny>> {
    let result = py
        .import("pytest._reporter")
        .and_then(|m| m.getattr("replacement"))
        .and_then(|f| f.call0());
    match result {
        Ok(reporter) if !reporter.is_none() => Some(reporter.unbind()),
        _ => None,
    }
}

/// Drive the replacement reporter's pytest_sessionstart (it owns the
/// session header in delegated mode).
pub fn reporter_sessionstart(py: Python<'_>, config: &crate::config::Config) {
    let result = (|| -> PyResult<()> {
        let session = make_session_proxy(py, config)?;
        py.import("pytest._reporter")?
            .getattr("sessionstart")?
            .call1((session,))?;
        Ok(())
    })();
    if let Err(err) = result {
        eprintln!("INTERNAL ERROR: {}", format_exception(py, &err));
    }
}

/// Drive the replacement reporter's pytest_collection_finish ("collected N
/// items" line). `numcollected` includes deselected items.
pub fn reporter_collection_finish(
    py: Python<'_>,
    config: &crate::config::Config,
    numcollected: usize,
) {
    let result = (|| -> PyResult<()> {
        let session = make_session_proxy(py, config)?;
        py.import("pytest._reporter")?
            .getattr("collection_finish")?
            .call1((session, numcollected))?;
        Ok(())
    })();
    if let Err(err) = result {
        eprintln!("INTERNAL ERROR: {}", format_exception(py, &err));
    }
}

/// Drive the replacement reporter's pytest_deselected (stats bookkeeping
/// behind the "X deselected" counts).
pub fn reporter_deselected(py: Python<'_>, items: &Bound<'_, PyAny>) {
    let result = (|| -> PyResult<()> {
        py.import("pytest._reporter")?
            .getattr("deselected")?
            .call1((items,))?;
        Ok(())
    })();
    if let Err(err) = result {
        eprintln!("INTERNAL ERROR: {}", format_exception(py, &err));
    }
}

/// The (file, lineno0, domain) location tuple for an item, as upstream
/// reports it to pytest_runtest_logstart/logfinish.
pub(crate) fn item_location<'py>(py: Python<'py>, item: &TestItem) -> PyResult<Bound<'py, PyAny>> {
    let file = item.nodeid.split("::").next().unwrap_or("");
    let domain = item
        .nodeid
        .split_once("::")
        .map(|(_, rest)| rest.replace("::", "."))
        .unwrap_or_default();
    Ok((file, item.lineno.saturating_sub(1), domain)
        .into_pyobject(py)?
        .into_any())
}

/// Drive the replacement reporter's pytest_runtest_logstart.
pub fn reporter_logstart(py: Python<'_>, item: &TestItem) {
    let result = (|| -> PyResult<()> {
        let location = item_location(py, item)?;
        py.import("pytest._reporter")?
            .getattr("logstart")?
            .call1((item.nodeid.as_str(), location))?;
        Ok(())
    })();
    if let Err(err) = result {
        eprintln!("INTERNAL ERROR: {}", format_exception(py, &err));
    }
}

/// Drive the replacement reporter's pytest_runtest_logreport.
pub fn reporter_logreport(py: Python<'_>, report: &Bound<'_, PyAny>) {
    let result = (|| -> PyResult<()> {
        reporter_logreport_fn(py)?.call1((report,))?;
        Ok(())
    })();
    if let Err(err) = result {
        eprintln!("INTERNAL ERROR: {}", format_exception(py, &err));
    }
}

/// Feed a report to the default reporter's stats (native mode only).
/// Populates terminalreporter.stats so conftest pytest_terminal_summary hooks
/// can access stats['passed'] etc. without the default reporter printing.
pub fn reporter_feed_default(py: Python<'_>, report: &Bound<'_, PyAny>) {
    let result = (|| -> PyResult<()> {
        reporter_feed_default_fn(py)?.call1((report,))?;
        Ok(())
    })();
    if let Err(err) = result {
        eprintln!("INTERNAL ERROR: {}", format_exception(py, &err));
    }
}

/// Ensure the default reporter has a trailing newline (plugins may
/// have written partial lines through the hook relay).
pub fn reporter_ensure_newline(py: Python<'_>) {
    let _ = py
        .import("pytest._reporter")
        .and_then(|m| m.getattr("ensure_newline"))
        .and_then(|f| f.call0());
}

/// Drain plugin-emitted reports (e.g., subtests) from the `_logreport_sink`.
/// These are reports that plugins logged via `ihook.pytest_runtest_logreport`
/// during normal (non-delegated) runs. Returns Rust `TestReport`s for the
/// engine to count and include in the session.
pub fn drain_plugin_reports(py: Python<'_>) -> Vec<crate::report::TestReport> {
    let result = (|| -> PyResult<Vec<crate::report::TestReport>> {
        let sink = py.import("_pytest.runner")?.getattr("_logreport_sink")?;
        let reports: Vec<Bound<'_, PyAny>> =
            sink.call_method0("drain_plugin_reports")?.extract()?;
        let mut out = Vec::new();
        for report in &reports {
            match crate::runner::report_from_proxy(py, report) {
                Ok(r) => out.push(r),
                Err(err) => {
                    eprintln!("INTERNAL ERROR: {}", format_exception(py, &err));
                }
            }
        }
        Ok(out)
    })();
    result.unwrap_or_default()
}

/// Subtest stat counts from the default reporter (populated by the
/// pytest-subtests plugin via the hook relay). Returns a map like
/// {"subtests passed": 3, "subtests failed": 2}.
pub fn reporter_subtest_stats(py: Python<'_>) -> std::collections::HashMap<String, usize> {
    py.import("pytest._reporter")
        .and_then(|m| m.getattr("subtest_stats"))
        .and_then(|f| f.call0())
        .and_then(|r| r.extract())
        .unwrap_or_default()
}

/// Drive the replacement reporter's pytest_runtest_logfinish.
pub fn reporter_logfinish(py: Python<'_>, item: &TestItem) {
    let result = (|| -> PyResult<()> {
        let location = item_location(py, item)?;
        py.import("pytest._reporter")?
            .getattr("logfinish")?
            .call1((item.nodeid.as_str(), location))?;
        Ok(())
    })();
    if let Err(err) = result {
        eprintln!("INTERNAL ERROR: {}", format_exception(py, &err));
    }
}

/// Feed a collection error to the replacement reporter as a failed
/// CollectReport (sugar prints these instantly; the base records stats).
pub fn reporter_collect_error(py: Python<'_>, nodeid: &str, longrepr: &str) {
    let result = (|| -> PyResult<()> {
        let kwargs = PyDict::new(py);
        kwargs.set_item("nodeid", nodeid)?;
        kwargs.set_item("outcome", "failed")?;
        kwargs.set_item("longrepr", longrepr)?;
        let file = nodeid.split("::").next().unwrap_or(nodeid);
        kwargs.set_item("location", (file, py.None(), file))?;
        kwargs.set_item("result", pyo3::types::PyList::empty(py))?;
        kwargs.set_item("sections", pyo3::types::PyList::empty(py))?;
        let report = py
            .import("_pytest.reports")?
            .getattr("CollectReport")?
            .call((), Some(&kwargs))?;
        // In a nested run, surface the failed collection to the HookRecorder
        // (getfailedcollections / assertoutcome read pytest_collectreport).
        crate::python::record_hook(
            py,
            "pytest_collectreport",
            &[("report", report.clone().unbind())],
        );
        py.import("pytest._reporter")?
            .getattr("collectreport")?
            .call1((report,))?;
        Ok(())
    })();
    if let Err(err) = result {
        eprintln!("INTERNAL ERROR: {}", format_exception(py, &err));
    }
}

/// In a nested run, surface a skipped collection (e.g. a DoctestModule that
/// could not import under --doctest-ignore-import-errors) to the HookRecorder
/// as a skipped CollectReport, matching upstream's collect-time Skipped.
/// No-op on the outer run.
pub fn record_collect_skip(py: Python<'_>, nodeid: &str, longrepr: &str) {
    if !crate::engine::inprocess::recording() {
        return;
    }
    let result = (|| -> PyResult<()> {
        let kwargs = PyDict::new(py);
        kwargs.set_item("nodeid", nodeid)?;
        kwargs.set_item("outcome", "skipped")?;
        kwargs.set_item("longrepr", longrepr)?;
        let file = nodeid.split("::").next().unwrap_or(nodeid);
        kwargs.set_item("location", (file, py.None(), file))?;
        kwargs.set_item("result", pyo3::types::PyList::empty(py))?;
        kwargs.set_item("sections", pyo3::types::PyList::empty(py))?;
        let report = py
            .import("_pytest.reports")?
            .getattr("CollectReport")?
            .call((), Some(&kwargs))?;
        crate::python::record_hook(py, "pytest_collectreport", &[("report", report.unbind())]);
        Ok(())
    })();
    if let Err(err) = result {
        eprintln!("INTERNAL ERROR: {}", format_exception(py, &err));
    }
}

/// End-of-run summaries through the replacement reporter (upstream's
/// pytest_terminal_summary / sessionfinish wrapper order).
pub fn reporter_finish(
    py: Python<'_>,
    config: &crate::config::Config,
    exitstatus: i32,
    shouldfail: Option<&str>,
) {
    let result = (|| -> PyResult<()> {
        let session = make_session_proxy(py, config)?;
        py.import("pytest._reporter")?
            .getattr("finish")?
            .call1((session, exitstatus, shouldfail))?;
        Ok(())
    })();
    if let Err(err) = result {
        eprintln!("INTERNAL ERROR: {}", format_exception(py, &err));
    }
}

/// Build the Python TestReport proxy handed to pytest_runtest_logreport
/// conftest hooks: a _pytest.reports.TestReport shim instance, whose
/// passed/failed/skipped flags and capstdout/capstderr/caplog properties
/// derive from the outcome and sections set here.
pub fn make_report_proxy(
    py: Python<'_>,
    report: &crate::report::TestReport,
    lineno: Option<u32>,
) -> PyResult<Py<PyAny>> {
    let (outcome, wasxfail) = match report.outcome {
        Outcome::Passed => ("passed", None),
        Outcome::Failed => ("failed", None),
        Outcome::Skipped => ("skipped", None),
        // pytest: expected failures are skipped/passed reports + .wasxfail.
        Outcome::XFailed => ("skipped", Some(report.longrepr.clone().unwrap_or_default())),
        Outcome::XPassed => ("passed", Some(String::new())),
    };
    let kwargs = PyDict::new(py);
    kwargs.set_item("nodeid", &report.nodeid)?;
    kwargs.set_item(
        "when",
        match report.phase {
            Phase::Setup => "setup",
            Phase::Call => "call",
            Phase::Teardown => "teardown",
        },
    )?;
    kwargs.set_item("outcome", outcome)?;
    kwargs.set_item("duration", report.duration.as_secs_f64())?;
    kwargs.set_item("longrepr", report.longrepr.as_deref())?;
    kwargs.set_item("sections", &report.sections)?;
    let file = report.nodeid.split("::").next().unwrap_or("");
    let domain = report
        .nodeid
        .split_once("::")
        .map(|(_, rest)| rest.replace("::", "."))
        .unwrap_or_default();
    kwargs.set_item(
        "location",
        (file, lineno.map(|l| l.saturating_sub(1)), domain),
    )?;
    kwargs.set_item("keywords", PyDict::new(py))?;
    kwargs.set_item("user_properties", pyo3::types::PyList::empty(py))?;
    if let Some(reason) = wasxfail {
        kwargs.set_item("wasxfail", reason)?;
    }
    let cls = reports_test_report_cls(py)?;
    let proxy = cls.call((), Some(&kwargs))?;
    // For subtest reports, reclassify the proxy as a SubtestReport so
    // Python hooks (pytest_report_teststatus) see the correct type and
    // call _sub_test_description() for the format word.
    if let Some(desc) = &report.subtest_desc {
        let sub_cls = py
            .import("pytest_subtests.plugin")
            .and_then(|m| m.getattr("SubTestReport"))
            .or_else(|_| {
                py.import("_pytest.subtests")
                    .and_then(|m| m.getattr("SubtestReport"))
            });
        if let Ok(sub_cls) = sub_cls {
            let is_passed = report.outcome == Outcome::Passed;
            let is_failed = report.outcome == Outcome::Failed;
            let is_skipped = report.outcome == Outcome::Skipped;
            let _ = proxy.setattr("passed", is_passed);
            let _ = proxy.setattr("failed", is_failed);
            let _ = proxy.setattr("skipped", is_skipped);
            let _ = proxy.setattr("__class__", &sub_cls);
            let desc_copy = desc.clone();
            let desc_fn =
                pyo3::types::PyCFunction::new_closure(py, None, None, move |_args, _kwargs| {
                    Ok::<String, PyErr>(desc_copy.clone())
                })?;
            let _ = proxy.setattr("_sub_test_description", &desc_fn);
            let _ = proxy.setattr("sub_test_description", desc_fn);
            let ctx_cls = py
                .import("pytest_subtests.plugin")
                .and_then(|m| m.getattr("SubTestContext"))
                .or_else(|_| {
                    py.import("_pytest.subtests")
                        .and_then(|m| m.getattr("SubtestContext"))
                });
            if let Ok(ctx_cls) = ctx_cls {
                let msg = desc
                    .strip_prefix('[')
                    .and_then(|s| s.find(']').map(|i| &s[..i]));
                let ctx_kwargs = PyDict::new(py);
                match msg {
                    Some(m) => ctx_kwargs.set_item("msg", m)?,
                    None => ctx_kwargs.set_item("msg", pyo3::types::PyNone::get(py))?,
                }
                ctx_kwargs.set_item("kwargs", PyDict::new(py))?;
                if let Ok(ctx) = ctx_cls.call((), Some(&ctx_kwargs)) {
                    let _ = proxy.setattr("context", ctx);
                }
            }
        }
    }
    Ok(proxy.unbind())
}

File: crates/pytest-rs-core/py/pytest/_reporter.py
"""Terminal-reporter delegation bridge.

pytest-rs renders terminal output natively (in Rust). Plugins like
pytest-sugar / pytest-pretty replace the 'terminalreporter' plugin during
pytest_configure; when that happens the engine suppresses its native output
and drives the replacement object through the same hook calls upstream
pluggy would make. The default TerminalReporter registered here is the
sacrificial stand-in those plugins unregister.
"""

from __future__ import annotations

import sys
import traceback
import warnings as _warnings
from typing import Any

from _pytest.terminal import TerminalReporter, WarningReport, _default_teststatus

from pytest._pluginmanager import _accepted_kwargs, instance_hook_impls, pluginmanager

_default: Any = None


def _trylast(func):
    """Mark a hook impl to dispatch last (pluggy trylast) without importing
    pytest here — the relay reads `pytest_impl`."""
    func.pytest_impl = {"trylast": True}
    return func


class _CoreHeader:
    """The rootdir/plugins header lines upstream's built-in
    pytest_report_header impls contribute (reporter-replacing plugins call
    config.hook.pytest_report_header and expect them)."""

    @staticmethod
    def pytest_report_header(config: Any) -> list[str]:
        lines = []
        try:
            verbose = int(getattr(config.option, "verbose", 0) or 0)
        except Exception:
            verbose = 0
        if verbose > 0:
            cache_dir = None
            try:
                cache_dir = config.getini("cache_dir")
            except Exception:
                pass
            lines.append(f"cachedir: {cache_dir or '.pytest_cache'}")
        rootpath = getattr(config, "rootpath", None)
        if rootpath is not None:
            lines.append(f"rootdir: {rootpath}")
        plugins = []
        try:
            from importlib.metadata import distributions

            registered_modules = {
                getattr(plugin, "__name__", None) for plugin in pluginmanager._plugins
            }
            for dist in distributions():
                for ep in dist.entry_points:
                    if ep.group != "pytest11":
                        continue
                    module = ep.value.split(":")[0].strip()
                    if module not in registered_modules:
                        continue
                    name = f"{dist.metadata['Name']}-{dist.version}"
                    if name.startswith("pytest-"):
                        name = name[7:]
                    if name not in plugins:
                        plugins.append(name)
        except Exception:
            pass
        if plugins:
            lines.append("plugins: {}".format(", ".join(sorted(plugins))))
        return lines


class _CoreTestStatus:
    """The (category, letter, word) default upstream's _pytest.runner /
    _pytest.skipping pytest_report_teststatus impls provide. Registered first
    so it dispatches last (the relay runs registered plugins LIFO), letting
    plugin impls win the firstresult. Without it, plugins that call
    config.hook.pytest_report_teststatus directly (pytest-bdd's
    gherkin-terminal-reporter) get None and crash unpacking it."""

    @staticmethod
    @_trylast
    def pytest_report_teststatus(report: Any, config: Any) -> Any:
        return _default_teststatus(report)


class _CorePyCollectMakeModule:
    """The default pytest_pycollect_makemodule impl upstream's _pytest.python
    provides: return a live Module collector node whose `.obj` is the imported
    module. Registered as a plain trylast impl so conftest hookwrappers
    (@pytest.hookimpl(wrapper=True)) surround it — a wrapper that mutates
    `mod.obj` (issue #205) thus mutates the real module the test functions read
    their globals from. The native engine fires this relay only when a conftest
    actually provides a makemodule hook, so default collection is unchanged."""

    @staticmethod
    @_trylast
    def pytest_pycollect_makemodule(module_path: Any, parent: Any) -> Any:
        from pytest._node import File

        node: Any = File(name=module_path.name, path=module_path, parent=parent)
        node.obj = getattr(parent, "_rs_module", None)
        node._rs_default_makemodule = True
        return node


class _CallInfo:
    """Minimal _pytest.runner.CallInfo stand-in for pytest_runtest_makereport
    consumers (pytest-bdd reads only item/report; others may read when)."""

    def __init__(self, when, excinfo=None):
        self.when = when
        self.excinfo = excinfo
        self.result = None
        self.start = 0.0
        self.stop = 0.0
        self.duration = 0.0


_makereport_result: Any = None


class _CoreMakeReport:
    """The plain (firstresult) pytest_runtest_makereport impl upstream's
    _pytest.runner provides: it returns the report the engine already built,
    which registered hookwrappers (pytest-bdd attaching .scenario) post-process
    before pytest_runtest_logreport. Registered first so it dispatches last."""

    @staticmethod
    @_trylast
    def pytest_runtest_makereport(item: Any, call: Any) -> Any:
        return _makereport_result


def run_makereport(report: Any, node: Any, when: str) -> Any:
    """Drive registered pytest_runtest_makereport hookwrappers over `report`
    so plugins can enrich it (pytest-bdd's .scenario) before it is logged."""
    global _makereport_result
    _makereport_result = report
    try:
        pluginmanager.hook.pytest_runtest_makereport(item=node, call=_CallInfo(when))
    finally:
        _makereport_result = None
    return report


def setup(config: Any) -> None:
    """Register the default terminalreporter (called before the engine
    fires pytest_configure into Python plugins)."""
    global _default
    if _default is not None:
        _default.__init__(config)
        return
    pluginmanager.register(_CoreHeader(), "_core_report_header")
    pluginmanager.register(_CoreTestStatus(), "_core_teststatus")
    pluginmanager.register(_CoreMakeReport(), "_core_makereport")
    pluginmanager.register(_CorePyCollectMakeModule(), "_core_pycollect_makemodule")
    _default = TerminalReporter(config)
    pluginmanager.register(_default, "terminalreporter")


def replacement() -> Any | None:
    """The non-default 'terminalreporter' plugin, or None when terminal
    output stays native."""
    # When setup() was never called (_default is None) — e.g. a -n worker,
    # whose only stdout is the IPC pipe — there is no pytest-rs-managed
    # default and no replacement to drive; the registered TerminalReporter
    # is the upstream default, which must stay silent (the master renders).
    if _default is None:
        return None
    reporter = pluginmanager.getplugin("terminalreporter")
    if reporter is None or reporter is _default:
        return None
    return reporter


def _call(obj: Any, name: str, /, **kwargs: Any) -> None:
    func = getattr(obj, name, None)
    if not callable(func):
        return
    try:
        func(**_accepted_kwargs(func, kwargs))
    except Exception:
        sys.stderr.write(f"INTERNALERROR in terminalreporter.{name}\n")
        traceback.print_exc()


# ---- engine-driven hook calls (no-ops unless a replacement registered) ----


def sessionstart(session: Any) -> None:
    """Drive the replacement's pytest_sessionstart: it owns the session
    header (the engine skipped its native one)."""
    reporter = replacement()
    if reporter is None:
        return
    reporter._session = session
    _call(reporter, "pytest_sessionstart", session=session)


def collection_finish(session: Any, numcollected: int) -> None:
    """The 'collected N items' line: the replacement's
    pytest_collection_finish (the base report_collect needs _numcollected,
    which native collection never fed it)."""
    reporter = replacement()
    if reporter is None:
        return
    reporter._session = session
    reporter._numcollected = numcollected
    _call(reporter, "pytest_collection_finish", session=session)


def deselected(items: list) -> None:
    reporter = replacement()
    if reporter is None:
        return
    _call(reporter, "pytest_deselected", items=items)


def logstart(nodeid: str, location: tuple) -> None:
    reporter = replacement()
    if reporter is not None:
        _call(reporter, "pytest_runtest_logstart", nodeid=nodeid, location=location)
    for impl in instance_hook_impls("pytest_runtest_logstart"):
        try:
            impl(nodeid=nodeid, location=location)
        except Exception:
            pass


def logreport(report: Any) -> None:
    reporter = replacement()
    if reporter is not None:
        _call(reporter, "pytest_runtest_logreport", report=report)
    for impl in instance_hook_impls("pytest_runtest_logreport"):
        try:
            impl(report=report)
        except Exception:
            pass


def logfinish(nodeid: str, location: tuple) -> None:
    reporter = replacement()
    if reporter is not None:
        _call(reporter, "pytest_runtest_logfinish", nodeid=nodeid, location=location)
    for impl in instance_hook_impls("pytest_runtest_logfinish"):
        try:
            impl(nodeid=nodeid, location=location)
        except Exception:
            pass


def collectreport(report: Any) -> None:
    reporter = replacement()
    if reporter is not None:
        _call(reporter, "pytest_collectreport", report=report)
    # Also dispatch to instance-registered plugins (e.g., relay plugin).
    for impl in instance_hook_impls("pytest_collectreport"):
        try:
            impl(report=report)
        except Exception:
            pass


def _feed_warnings(reporter: Any) -> None:
    """Mirror captured warnings into reporter.stats['warnings'] (upstream
    feeds them live via pytest_warning_recorded)."""
    try:
        from pytest import _wcapture

        stats = getattr(reporter, "stats", None)
        if stats is None or stats.get("warnings"):
            return
        entries = []
        for warning in _wcapture.captured:
            message = _warnings.formatwarning(
                warning["message"],  # type: ignore[arg-type]
                warning["category"],  # type: ignore[arg-type]
                warning["filename"],  # type: ignore[arg-type]
                warning["lineno"],  # type: ignore[arg-type]
            )
            entries.append(
                WarningReport(
                    message,
                    nodeid=warning["test"],
                    fslocation=(warning["filename"], warning["lineno"]),
                )
            )
        if entries:
            stats["warnings"] = entries
    except Exception:
        pass


_fed_counts: dict[str, int] = {}


def _track_delegated_report(report: Any) -> None:
    """Account for a report the default reporter received via the shim PM
    during a delegated protocol.  Incrementing _fed_counts prevents
    subtest_stats() from returning it as an 'extra' count (the engine
    already tracks it in session.reports)."""
    if _default is None:
        return
    try:
        category, _, _ = _default._gettestkindstatus(report)
        _fed_counts[category] = _fed_counts.get(category, 0) + 1
    except Exception:
        pass


def feed_default(report: Any) -> None:
    """Feed a report to the default reporter's stats without printing terminal
    output. Used in native mode so conftest pytest_terminal_summary hooks can
    access terminalreporter.stats['passed'] etc."""
    if _default is None:
        return
    try:
        category, _, _ = _default._gettestkindstatus(report)
        _default._add_stats(category, [report])
        _fed_counts[category] = _fed_counts.get(category, 0) + 1
    except Exception:
        pass


def ensure_newline() -> None:
    """Ensure the default reporter has a trailing newline (plugins may
    have written partial lines through the relay)."""
    if _default is not None:
        _default.ensure_newline()


def subtest_stats() -> dict[str, int]:
    """Return plugin-driven stat counts from the default reporter that
    the native engine does not see (reports emitted through the hook
    relay, not through session.reports). Subtracts counts fed by the
    engine via feed_default to avoid double-counting."""
    if _default is None:
        return {}
    result = {}
    for key, items in _default.stats.items():
        if not items or key in ("", "warnings", "deselected", "error"):
            continue
        extra = len(items) - _fed_counts.get(key, 0)
        if extra > 0:
            result[key] = extra
    return result


def finish(session: Any, exitstatus: int, shouldfail: str | None = None) -> None:
    """End-of-run summaries, in upstream's pytest_terminal_summary /
    pytest_sessionfinish wrapper order (the sessionfinish wrapper closes the
    progress line first, then summaries, then the stats line)."""
    # Fire pytest_sessionfinish on instance-registered plugins (e.g., relay plugin).
    for impl in instance_hook_impls("pytest_sessionfinish"):
        try:
            impl(session=session, exitstatus=exitstatus)
        except Exception:
            pass
    reporter = replacement()
    if reporter is None:
        return
    _feed_warnings(reporter)
    try:
        reporter._tw.line("")
    except Exception:
        pass
    _call(reporter, "summary_errors")
    _call(reporter, "summary_failures")
    _call(reporter, "summary_xfailures")
    _call(reporter, "summary_warnings")
    _call(reporter, "summary_passes")
    _call(reporter, "summary_xpasses")
    # Other plugins' pytest_terminal_summary impls (the replacement itself
    # defines none — upstream's is a hookwrapper this sequence replays).
    try:
        pluginmanager.hook.pytest_terminal_summary(
            terminalreporter=reporter, exitstatus=exitstatus, config=reporter.config
        )
    except Exception:
        traceback.print_exc()
    _call(reporter, "short_test_summary")
    _call(reporter, "summary_warnings")
    if shouldfail:
        _call(reporter, "write_sep", sep="!", title=str(shouldfail), red=True)
    _call(reporter, "summary_stats")
